handlers/generic.go

```diff
@@ -12,179 +12,97 @@
 //    See the License for the specific language governing permissions and
 //    limitations under the License.
 
-package parser
+package handlers
 
 import (
-	"archive/tar"
 	"bytes"
-	"compress/gzip"
-	"crypto/sha256"
-	"encoding/hex"
 	"io"
 	"io/ioutil"
+	"os"
 	"path/filepath"
 	"strings"
 
-	"github.com/mendersoftware/mender-artifact/metadata"
+	"github.com/mendersoftware/mender-artifact/artifact"
 	"github.com/pkg/errors"
 )
 
-type GenericParser struct {
-	metadata   metadata.Metadata
-	updates    map[string]UpdateFile
-	typeParsed string
+type Generic struct {
+	updateType string
+	files      map[string](*artifact.DataFile)
 }
 
-func (rp *GenericParser) GetUpdateType() *metadata.UpdateType {
-	return &metadata.UpdateType{Type: rp.typeParsed}
+func NewGeneric(t string) *Generic {
+	return &Generic{
+		updateType: t,
+		files:      make(map[string](*artifact.DataFile)),
+	}
 }
 
-func (rp *GenericParser) GetUpdateFiles() map[string]UpdateFile {
-	return rp.updates
+func (g *Generic) GetUpdateFiles() [](*artifact.DataFile) {
+	list := make([](*artifact.DataFile), len(g.files))
+	i := 0
+	for _, f := range g.files {
+		list[i] = f
+		i++
+	}
+	return list
 }
 
-func (rp *GenericParser) GetMetadata() *metadata.Metadata {
-	return &rp.metadata
+func (g *Generic) GetType() string {
+	return g.updateType
 }
 
-func parseFiles(tr *tar.Reader, uFiles map[string]UpdateFile) error {
-	files := new(metadata.Files)
-	if _, err := io.Copy(files, tr); err != nil {
-		return errors.Wrapf(err, "parser: error reading files")
-	}
-	for _, file := range files.FileList {
-		uFiles[filepath.Base(file)] = UpdateFile{Name: file}
-	}
+// Copy is implemented only to satisfy Installer interface.
+// Generic parser is not supposed to be copied.
+func (g *Generic) Copy() artifact.Installer {
 	return nil
```

Comment:
I think this is OK for now. Maybe we can think of some smart compile time assert later.